maketeams3.py

```diff
@@ -258,60 +313,219 @@ def testSwap(teama, playera, teamb, playerb, board):
+        return False
+
+    return True
+
+
+def get_swaps(teams):
+    num_boards = len(teams[0].boards)
+    boards = [[team.boards[i] for team in teams] for i in range(num_boards)]
+    swaps = [[swap for swap in combinations(board, 2) if is_neutral_swap(swap)] for board in boards]
+    return [swap for board_swaps in swaps for swap in board_swaps]
+
+
+def rating_variance_improvement(league_mean, n_boards, swap):
+    def score(a, b):
+        return variance(league_mean, [a, b])
+
+    pa, pb = swap
+    a_mean = pa.team.getMean()
+    b_mean = pb.team.getMean()
+    initial_score = score(a_mean, b_mean)
+
+    # calculating change in mean if we swapped the players.
+    rating_diff = pb.rating - pa.rating
+    a_mean = a_mean + rating_diff/n_boards
+    b_mean = b_mean - rating_diff/n_boards
+    new_score = score(a_mean, b_mean)
+
+    # lower is better
+    return new_score - initial_score
+
+
+def get_best_swap(swaps, fun):
+    best_swap = min(swaps, key=fun)
+    return best_swap, fun(best_swap)
+
+
+def perform_swap(swap):
+    pa, pb = swap
+    ta = pa.team
+    tb = pb.team
+    board = pa.board
+    ta.changeBoard(board, pb)
+    tb.changeBoard(board, pa)
+
+
+def update_swaps(swaps, swap_performed, teams):
+    pa, pb = swap_performed
+    affected_players = pa.team.boards + pb.team.boards
+    # remove all swaps involving players affected by the swap.
+    swaps = [swap for swap in swaps
+             if not (swap[0] in affected_players or swap[1] in affected_players)]
+
+    # find new neutral swaps involving the players affected by swap.
+    for player in affected_players:
+        board = player.board
+        players_on_board = [team.boards[board] for team in teams
+                            if not team.boards[board] in affected_players]
+        swaps.extend([(player, p) for p in players_on_board
+                      if is_neutral_swap((player, p))])
+
+    swaps.extend([swap for swap in zip(pa.team.boards, pb.team.boards)
+                  if is_neutral_swap(swap)])
+
+    return swaps
+
+
+def reduce_variance(teams):
+    # players = sum([team.boards for team in teams], [])
+
+    league_mean = sum([team.getMean() for team in teams]) / len(teams)
+    n_boards = len(teams[0].boards)
+
+    swaps = get_swaps(teams)
+
+    eval_fun = partial(rating_variance_improvement, league_mean, n_boards)
+    best_swap, swap_value = get_best_swap(swaps, eval_fun)
+
+    # infinite loop possibility here?
+    i = 0
+    max_iterations = 200
+    epsilon = 0.0000001
+    while swap_value <= -epsilon and i < max_iterations:
+        # variance = team_rating_variance(teams, league_mean)
+        # updatePref(players, teams)
+        # score = total_happiness(teams)
+        # print()
+        # print("i: ", i)
+        # print("variance: ", variance)
+        # print("score: ", score)
+        # print("swap_value: ", swap_value)
+        # print("best_swap: ", best_swap)
```

Comment:
The best_swap will always decrease the rating variance, right? If so, then I don't see how it would be an infinite loop. At some point you'll hit an optimal configuration.  I'd be interested in seeing the graph of rating variance improvement for each iteration.  That way we can make some decisions about how long we want to run it for.

Reply:
Yeah, this doesn't happen. It should! but it doesn't.  I guess there's a problem somewhere but I didn't get into that.

Reply:
Oh, I just remembered why I didn't bother with this.  I assumed this must be fp math junk.  When two players are identical in rating then the effect of the swap should be zero.  But maybe the math I'm doing never returns zero.  zero will always be an extremely small number.  I don't know that this is the case. I just assumed it.